sycl/include/syclcompat/group_utils.hpp

```diff
@@ -0,0 +1,1269 @@
+  ///
+  ///   0, 1, 2, 3, 4, 5, 0, 0, ..., 0, 0, 0, 0.
+  ///
+  /// The striped order \p output will be:
+  ///
+  ///   0, 4, 8, 12, 16, 0, 0, 0, ..., 0, 0, 0, 0.
+  ///
+  /// \tparam ItemT The sycl::nd_item index space class.
+  /// \tparam OutputIteratorT The random-access iterator type for \p output
+  /// iterator.
+  /// \param item The work-item identifier.
+  /// \param input The input data of each work-item.
+  /// \param data The data to store.
+  /// \param valid_items Number of valid items to load
+  template <typename ItemT, typename OutputIteratorT>
+  __syclcompat_inline__ void
+  store(const ItemT &item, OutputIteratorT output_iter,
+        T (&data)[ElementsPerWorkItem], size_t valid_items) {
+    if constexpr (StoreAlgorithm == group_store_algorithm::blocked) {
+      store_direct_blocked<T, ElementsPerWorkItem, OutputIteratorT, ItemT>(
+          item, output_iter, data, valid_items);
+    } else if constexpr (StoreAlgorithm == group_store_algorithm::striped) {
+      store_direct_striped<T, ElementsPerWorkItem, OutputIteratorT, ItemT>(
+          item, output_iter, data, valid_items);
+    }
+  }
+};
+
+/// The work-group wide shuffle operations that allow work-items to exchange
+/// data elements with other work-items within the same work-group.
+///
+/// \tparam T The type of the data elements.
+/// \tparam group_dim_0 The first dimension size of the work-group.
+/// \tparam group_dim_1 The second dimension size of the work-group.
+/// \tparam group_dim_2 The third dimension size of the work-group.
+template <typename T, int group_dim_0, int group_dim_1 = 1, int group_dim_2 = 1>
+class group_shuffle {
+  T *_local_memory = nullptr;
+  static constexpr size_t group_work_items =
+      group_dim_0 * group_dim_1 * group_dim_2;
+
+public:
+  static constexpr size_t get_local_memory_size(size_t work_group_size) {
+    return sizeof(T) * work_group_size;
+  }
+  group_shuffle(uint8_t *local_memory) : _local_memory((T *)local_memory) {}
+
+  /// Selects a value from a work-item at a given distance in the work-group
+  /// and stores the value in the output.
+  ///
+  /// \tparam ItemT The work-item identifier type.
+  /// \param item The work-item identifier.
+  /// \param input The input from the calling work-item.
+  /// \param output The output where the selected data will be stored.
+  /// \param distance The distance of work-items to look ahead or behind in the
+  /// work-group.
+  template <typename ItemT>
+  __syclcompat_inline__ void select(const ItemT &item, T input, T &output,
+                                    int distance = 1) {
+    auto g = item.get_group();
+    size_t id = g.get_local_linear_id();
+    _local_memory[id] = input;
+
+    sycl::group_barrier(g, sycl::memory_scope::work_group);
+
+    const int target_id = static_cast<int>(id) + distance;
+    if ((target_id >= 0) && (target_id < group_work_items)) {
+      output = _local_memory[static_cast<size_t>(target_id)];
```

Comment:
Why not use existing performance tuned algorithms that are already standard sycl. e.g. something like
```suggestion
    sycl::select_from_group(g, distance, id);
```
See:

https://github.com/intel/llvm/blob/68cacbfed4f54beb5ce61b9963aa6ddca61eebb5/sycl/include/sycl/group_algorithm.hpp#L596

Reply:
OK, maybe I am wrong because CUB also seems to use shared memory. https://github.com/NVIDIA/cub/blob/0fc3c3701632a4be906765b73be20a9ad0da603d/cub/block/block_shuffle.cuh#L31

Apologies, I am probably misunderstanding something.